### PR TITLE
minor enhancements: return technical metadata, allow job ID specification for testing

### LINKED ISSUE #24: speech-to-text worker sends back some basic technical metadata in the body of the done message it queues
this is akin to the info we log in the DOR event log about the abbyy version used for text extraction in the `ocrWF`, see:
* https://github.com/sul-dlss/common-accessioning/blob/ffb7f37ebc6899ca2699097868f1940d8f54b714/lib/dor/text_extraction/abbyy/file_watcher.rb#L85-L86
* https://github.com/sul-dlss/common-accessioning/blob/ffb7f37ebc6899ca2699097868f1940d8f54b714/lib/dor/text_extraction/abbyy/results.rb#L128-L145

### PR DESCRIPTION
**Pull request by jmartin-sul**
two small improvements, one functional for actual operation, one that helped me in my testing:
* technical metadata about whisper run passed back in done message when extraction completes
* allow manual specification of `--job_id` when `--create` mode is used

closes #24 (well, maybe, see inline comment/question)

example `extraction_technical_metadata` field value as it would come back to common-accessioning's watcher daemon, for copying to the dor-services-app event log entry about the speech-to-text attempt:

```
{\"speech_to_text_extraction_program\": {\"name\": \"whisper\", \"version\": \"20240930\"}, \"effective_options\": {\"fp16\": false}, \"effective_writer_options\": {\"highlight_words\": null, \"max_line_count\": null, \"max_line_width\": null, \"max_words_per_line\": null}, \"effective_model_name\": \"small\"}
```

result when the JSON string is parsed to a more readable ruby hash:
```ruby
{"speech_to_text_extraction_program"=>{"name"=>"whisper", "version"=>"20240930"},
 "effective_options"=>{"fp16"=>false},
 "effective_writer_options"=>
  {"highlight_words"=>nil, "max_line_count"=>nil, "max_line_width"=>nil, "max_words_per_line"=>nil},
 "effective_model_name"=>"small"}
```

i was able to see the information come back in the "done" SQS message handled by the watcher i was testing for https://github.com/sul-dlss/common-accessioning/pull/1388

### CODE REVIEW

File: speech_to_text.py [hunk 1]
```diff
@@ -114,6 +114,13 @@ def run_whisper(job):
         writer(result, media_file, writer_options)
 
     job["finished"] = now()
+    tech_md_dict = {
+        "speech_to_text_extraction_program": {"name": "whisper", "version": whisper.version.__version__},
+        "effective_options": options,
+        "effective_writer_options": writer_options,
+        "effective_model_name": model_name
```
Comment (jmartin-sul) [inline speech_to_text.py:121]: another thing i considered adding, but left off as possibly too verbose and not useful enough: `"effective_model": str(model)`

this produced output like:
```
Whisper(\n  (encoder): AudioEncoder(\n    (conv1): Conv1d(80, 768, kernel_size=(3,), stride=(1,), padding=(1,))\n    (conv2): Conv1d(768, 768, kernel_size=(3,), stride=(2,), padding=(1,))\n    (blocks): ModuleList(\n      (0-11): 12 x ResidualAttentionBlock(\n        (attn): MultiHeadAttention(\n          (query): Linear(in_features=768, out_features=768, bias=True)\n          (key): Linear(in_features=768, out_features=768, bias=False)\n          (value): Linear(in_features=768, out_features=768, bias=True)\n          (out): Linear(in_features=768, out_features=768, bias=True)\n        )\n        (attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n        (mlp): Sequential(\n          (0): Linear(in_features=768, out_features=3072, bias=True)\n          (1): GELU(approximate='none')\n          (2): Linear(in_features=3072, out_features=768, bias=True)\n        )\n        (mlp_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n      )\n    )\n    (ln_post): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n  )\n  (decoder): TextDecoder(\n    (token_embedding): Embedding(51865, 768)\n    (blocks): ModuleList(\n      (0-11): 12 x ResidualAttentionBlock(\n        (attn): MultiHeadAttention(\n          (query): Linear(in_features=768, out_features=768, bias=True)\n          (key): Linear(in_features=768, out_features=768, bias=False)\n          (value): Linear(in_features=768, out_features=768, bias=True)\n          (out): Linear(in_features=768, out_features=768, bias=True)\n        )\n        (attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n        (cross_attn): MultiHeadAttention(\n          (query): Linear(in_features=768, out_features=768, bias=True)\n          (key): Linear(in_features=768, out_features=768, bias=False)\n          (value): Linear(in_features=768, out_features=768, bias=True)\n          (out): Linear(in_features=768, out_features=768, bias=True)\n        )\n        (cross_attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n        (mlp): Sequential(\n          (0): Linear(in_features=768, out_features=3072, bias=True)\n          (1): GELU(approximate='none')\n          (2): Linear(in_features=3072, out_features=768, bias=True)\n        )\n        (mlp_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n      )\n    )\n    (ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)\n  )\n)
```

the non-stringified and slightly more readable version i got when playing around with this in parts on the python REPL):
```
>>> model
Whisper(
  (encoder): AudioEncoder(
    (conv1): Conv1d(80, 768, kernel_size=(3,), stride=(1,), padding=(1,))
    (conv2): Conv1d(768, 768, kernel_size=(3,), stride=(2,), padding=(1,))
    (blocks): ModuleList(
      (0-11): 12 x ResidualAttentionBlock(
        (attn): MultiHeadAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=False)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (out): Linear(in_features=768, out_features=768, bias=True)
        )
        (attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): Sequential(
          (0): Linear(in_features=768, out_features=3072, bias=True)
          (1): GELU(approximate='none')
          (2): Linear(in_features=3072, out_features=768, bias=True)
        )
        (mlp_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      )
    )
    (ln_post): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (decoder): TextDecoder(
    (token_embedding): Embedding(51865, 768)
    (blocks): ModuleList(
      (0-11): 12 x ResidualAttentionBlock(
        (attn): MultiHeadAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=False)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (out): Linear(in_features=768, out_features=768, bias=True)
        )
        (attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (cross_attn): MultiHeadAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=False)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (out): Linear(in_features=768, out_features=768, bias=True)
        )
        (cross_attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): Sequential(
          (0): Linear(in_features=768, out_features=3072, bias=True)
          (1): GELU(approximate='none')
          (2): Linear(in_features=3072, out_features=768, bias=True)
        )
        (mlp_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      )
    )
    (ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
)
```

side-note looks oddly LISP-y to me (says someone who last wrote LISP in a college class 20+ years ago.

**Comment by andrewjbtw** [comment]
@jmartin-sul is the technical metadata about the whisper run meant to go into the Events?

**Comment by jmartin-sul** [comment]
> @jmartin-sul is the technical metadata about the whisper run meant to go into the Events?

yup.  this PR doesn't enforce that, since it's up to the consumer of the message to do something with the info.  but https://github.com/sul-dlss/common-accessioning/pull/1388 will put it in the DOR event log, since it logs the entire payload of the done message (i.e. all the stuff we stick in the message, but not the surrounding AWS/SQS message metadata).

File: speech_to_text.py (hunk 1 above)
Comment (andrewjbtw) [reply speech_to_text.py:121]: Possibly worth a post-standup discussion. I would not find this level of detail useful to record, but perhaps someone more familiar with whisper would see some data in there they'd want?

**Review by edsu — APPROVED** [review]
Sorry I didn't get to this sooner. I like the idea of passing back the options that were actually used to run Whisper. If we need more information about the result we can add it later. I do think that sharing information that isn't immediately needed can be helpful sometimes, so that people can discover it. But it can also be distracting from the information that is actually needed.

**Comment by jmartin-sul** [comment]
> Sorry I didn't get to this sooner. I like the idea of passing back the options that were actually used to run Whisper. If we need more information about the result we can add it later. I do think that sharing information that isn't immediately needed can be helpful sometimes, so that people can discover it. But it can also be distracting from the information that is actually needed.

no worries!  and agreed.

**Pull request merged**